src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD113CheckForSystemIAsyncDisposableAnalyzer.cs

```diff
@@ -0,0 +1,104 @@
+            defaultSeverity: DiagnosticSeverity.Info,
+            isEnabledByDefault: true);
+
+        /// <inheritdoc />
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Descriptor);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze);
+
+            context.RegisterCompilationStartAction(startCompilation =>
+            {
+                INamedTypeSymbol? vsThreadingAsyncDisposableType = startCompilation.Compilation.GetTypeByMetadataName(Types.IAsyncDisposable.FullName);
+                INamedTypeSymbol? bclAsyncDisposableType = startCompilation.Compilation.GetTypeByMetadataName(Types.BclAsyncDisposable.FullName);
+                if (vsThreadingAsyncDisposableType is object)
+                {
+                    startCompilation.RegisterOperationAction(Utils.DebuggableWrapper(c => this.AnalyzeTypeCheck(c, vsThreadingAsyncDisposableType, bclAsyncDisposableType)), OperationKind.IsType);
+                    startCompilation.RegisterOperationAction(Utils.DebuggableWrapper(c => this.AnalyzeTypeCheck(c, vsThreadingAsyncDisposableType, bclAsyncDisposableType)), OperationKind.IsPattern);
+                    startCompilation.RegisterOperationAction(Utils.DebuggableWrapper(c => this.AnalyzeTypeCheck(c, vsThreadingAsyncDisposableType, bclAsyncDisposableType)), OperationKind.Conversion);
+                }
+            });
+        }
+
+        private void AnalyzeTypeCheck(OperationAnalysisContext context, INamedTypeSymbol vsThreadingAsyncDisposableType, INamedTypeSymbol bclAsyncDisposableType)
+        {
+            switch (context.Operation)
+            {
+                case IIsTypeOperation { TypeOperand: { } operand }:
+                    ConsiderTypeCheck(context, operand, vsThreadingAsyncDisposableType, bclAsyncDisposableType);
+                    break;
+                case IIsPatternOperation { Pattern: IDeclarationPatternOperation { DeclaredSymbol: ILocalSymbol { Type: { } operand } } }:
+                    ConsiderTypeCheck(context, operand, vsThreadingAsyncDisposableType, bclAsyncDisposableType);
+                    break;
+                case IConversionOperation { Type: { } operand }:
+                    ConsiderTypeCheck(context, operand, vsThreadingAsyncDisposableType, bclAsyncDisposableType);
+                    break;
+            }
+
+            static void ConsiderTypeCheck(OperationAnalysisContext context, ITypeSymbol operand, INamedTypeSymbol vsThreadingAsyncDisposableType, INamedTypeSymbol bclAsyncDisposableType)
+            {
+                if (Equals(vsThreadingAsyncDisposableType, operand))
+                {
+                    // If System.IAsyncDisposable isn't even defined, they can't have the check we would look for.
+                    if (bclAsyncDisposableType is object)
```

Comment:
Nit: This comment is a bit difficult to parse since it is actually describing the not condition and not what we are doing in the if block. Maybe something like "only check further if BCL type is defined, otherwise just report the missing check"